### DISCUSSION TITLE
Instructions on installing sentry-native SDK on linux (ubuntu 22.04)

### DISCUSSION THREAD

**Discussion by tian-yu-ocado** [getsentry/sentry-native#866] [Q&A] [answered]
Hi there,

We would like to install sentry-native SDK on docker containers with ubuntu 22.04. Is there any specific instructions on how to do that? i.e. A step by step guide on how to: 1. Properly download and install the dependencies; 2. How to configure the sentry-native cmake arguments to install to proper system directories such as `/usr/local/lib`?

Also, is there an existing `.deb` package for sentry-native we could directly download and install? If not is there a plan to create one? I would imagine that would be useful for a broader user spectrum, especially those who are not familiar with building/installing packages from source code.

**Comment by supervacuus** [comment, marked as answer]
Hi @tian-yu-ocado!

Building and installing the Sentry Native SDK for `docker` is not that different from doing so on a regular Linux host. For `docker` specifically, our users choose the `breakpad` backend (which is also the default on Linux) because this way, you do not need to maintain a separate process beyond your executables lifetime (as you would have to with `crashpad` with processes crashes via a separate executable which outlives the crashed process).

The process of installing and building the Native SDK is explained here: https://docs.sentry.io/platforms/native/#install. Download the `sentry-native.zip` since it will contain vendored source dependencies.

After downloading and extracting the ZIP file, specifically for Ubuntu 22.04, I would recommend installing the following packages first:

`sudo apt install cmake build-essential zlib1g-dev libcurl4-openssl-dev`

Whereas the first two are necessary tools for the build, the latter are dependencies. You can use any `libcurl4-*-dev` package, but typically since other programs require it as well, you will use the one that defers the `TLS` implementation to `openssl`.

A basic build on Linux requires the same steps as you can gather from the install documentation above:

```
# 1. Configure the build to use the breakpad backend (while it is the default, always prefer to be explicit) and make it a Release build (in this case, with debug-info generated as well, but you could also use `Release`). This command will tell you if you missed any dependencies.
cmake -B build -DSENTRY_BACKEND=breakpad -DCMAKE_BUILD_TYPE=RelWithDebInfo

# 2. After successful configuration, build the Native SDK. Don't use `--parallel` if you build on a low-resource VM.
cmake --build build --parallel

# 3. Install the build results in /usr/local (this command requires permission to write to /usr/local)
cmake --install build --prefix /usr/local
```

When using `docker`, you'd probably want to set up a [multi-stage build](https://docs.docker.com/build/building/multi-stage/) and do all of the Native SDK steps in a separate stage and then only copy the install directory to your final stage. But this goes beyond where we can help you because it requires insight into your specific needs for building and deploying a docker image.

No Debian packages are available for `sentry-native`, nor do we plan to provide any soon. The community provides build-time packaging for `vcpkg` and `conan`, though, and someone maintains packages for `nix` (providing only a single configuration using the system `breakpad`). The reasoning is that the Native SDK provides many different build-time configurations to support the extensive range of different use cases our users are exposed to (resulting in a cross-product of Debian packages that we'd have to support). 

But we are always interested in input from the community. Particularly: which use-case do you have that requires using the Native SDK without you being familiar with building and installing libraries from source? Do you not use the Native SDK in a C/C++ project? Do you have a specific environment that complicates the usage of `conan` or `vcpkg`? Please let us know, thx!

**Reply by tian-yu-ocado** [reply]
Hi @supervacuus:

Thank you for your kind reply and detailed answer! I followed your steps (minus the specification of sentry_backend and build_type), and it worked perfectly. I also used git clone --recurse-submodules to properly get the source code instead of downloading the .zip and unzip it. I guess it was a bit difficult for me in the beginning as I am not familiar with the steps of build dependencies from source code. I would suggest to have some step by step guides (exactly as what you put down in this answer) as of how we could properly: 1. Download the source file 2. use make/cmake to build the source code 3. install the libraries to desired locations. It would also be helpful to those who does not know how cmake works :)
In terms of your questions: I am using the Native SDK with our c++ project and have some knowledge about building libraries from source using cmake, but I guess what tricked me from the beginning was that I did not download the `breakpad` source code properly so the build of sentry-native was not successful, and I was circling around how to build and install `breakpad` properly such that sentry-native could find it and build. But once I have the source tree of `sentry-native` properly downloaded it worked pretty smoothly.
In terms of `conan` and `vcpkg` I do find the tools themselves are complicated to use combined with `cmake`, so I tend to avoid them :)

**Reply by supervacuus** [reply]
Glad I could be of help @tian-yu-ocado!

> I would suggest to have some step by step guides (exactly as what you put down in this answer) as of how we could properly: 1. Download the source file 2. use make/cmake to build the source code 3. install the libraries to desired locations. It would also be helpful to those who does not know how cmake works :)

I think these 3 points are described with the same level of detail in the install section of our docs: https://docs.sentry.io/platforms/native/#install.

>  In terms of your questions: I am using the Native SDK with our c++ project and have some knowledge about building libraries from source using cmake, but I guess what tricked me from the beginning was that I did not download the breakpad source code properly so the build of sentry-native was not successful, and I was circling around how to build and install breakpad properly such that sentry-native could find it and build. But once I have the source tree of sentry-native properly downloaded it worked pretty smoothly.

Okay, that makes sense. This is why we guide users in the install docs to download the release packages instead of setting up the GitHub repo and submodules themselves. Once they know how it works, using submodules isn't really an obstacle anymore, but it helps with jumping into the first setup.

**Comment by misje** [comment]
I am providing deb packages [here](https://github.com/misje/sentry-native-deb/releases) for Ubuntu 22.04 (Debian Bookworm) and newer.
